File: Practice!/Week2/LinkedListPrac.py
# Practice1
# 단방향 연결 리스트에서 중복 데이터를 찾아 삭제하세요.

# 입출력 예시)
# 입력 연결 리스트: 1, 3, 3, 1, 4, 2, 4, 2
# 결과 연결 리스트: 1, 3, 4, 2
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def add_data(self, data):
        if not self.head:
            self.head = Node(data)
        else:
            cur = self.head
            while cur.next:
                cur = cur.next
            cur.next = Node(data)

    def find_data(self, data):
        cur = self.head
        while cur:
            if cur.data == data:
                return True
            cur = cur.next
        return False

    def show_data(self):
        cur = self.head
        while cur:
            print(cur.data, end=" ")
            cur = cur.next
        print()


def remove_dup(list_before):
    #여기에 코드 작성
    return null


# Test code
linked_list = LinkedList()
linked_list.add_data(1)
linked_list.add_data(3)
linked_list.add_data(3)
linked_list.add_data(1)
linked_list.add_data(4)
linked_list.add_data(2)
linked_list.add_data(4)
linked_list.add_data(2)

print("Before removing duplicates:")
linked_list.show_data()

result = remove_dup(linked_list)



# Practice2
# Palindrome 연결 리스트
# 추가 자료구조 없이 연결 리스트만으로 풀어보기 Practice1에서 만든 단일연결리스트 사용(내장 LinkedList말고)
# Palindrome: 앞으로 읽어도 뒤로 읽어도 같은 문자열
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def add_data(self, data):
        if not self.head:
            self.head = Node(data)
        else:
            cur = self.head
            while cur.next:
                cur = cur.next
            cur.next = Node(data)

    def find_data(self, data):
        cur = self.head
        while cur:
            if cur.data == data:
                return True
            cur = cur.next
        return False

    def show_data(self):
        cur = self.head
        while cur:
            print(cur.data, end=" ")
            cur = cur.next
        print()

    def is_palindrome(self):
        //여기에 코드 작성
        return false

    


# Test code
linked_list1 = LinkedList()
linked_list1.add_data(1)
linked_list1.add_data(3)
linked_list1.add_data(5)
linked_list1.add_data(3)
linked_list1.add_data(1)
print(linked_list1.is_palindrome())  # Output: True

linked_list2 = LinkedList()
linked_list2.add_data(3)
linked_list2.add_data(5)
linked_list2.add_data(5)
linked_list2.add_data(3)
print(linked_list2.is_palindrome())  # Output: True

linked_list3 = LinkedList()
linked_list3.add_data(1)
linked_list3.add_data(3)
linked_list3.add_data(5)
linked_list3.add_data(1)
print(linked_list3.is_palindrome())  # Output: False


print("After removing duplicates:")
result.show_data()




# Practice3
# 연결 리스트 부분 뒤집기
# 주어진 연결 리스트에서 시작 위치부터 끝 위치 사이의 노드들을 뒤집으시오.

# 입력 예시)
# 입력 연결 리스트: 1, 2, 3, 4, 5
# 시작 위치: 2
# 끝 위치: 4
# (처음 위치는 1부터라고 가정)
# 결과 연결 리스트: 1, 4, 3, 2, 5
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def add_data(self, data):
        if not self.head:
            self.head = Node(data)
        else:
            cur = self.head
            while cur.next:
                cur = cur.next
            cur.next = Node(data)

    def find_data(self, data):
        cur = self.head
        while cur:
            if cur.data == data:
                return True
            cur = cur.next
        return False

    def show_data(self):
        cur = self.head
        while cur:
            print(cur.data, end=" ")
            cur = cur.next
        print()


def reverse_list(linked_list, left, right):
    //여기에 코드 작성 
    return linked_list

# Test cases
linked_list1 = LinkedList()
linked_list1.add_data(1)
linked_list1.add_data(2)
linked_list1.add_data(3)
linked_list1.add_data(4)
linked_list1.add_data(5)
linked_list1.show_data()

linked_list1 = reverse_list(linked_list1, 2, 4)
linked_list1.show_data()

linked_list2 = LinkedList()
linked_list2.add_data(1)
linked_list2.add_data(2)
linked_list2.add_data(3)
linked_list2.add_data(4)
linked_list2.add_data(5)
linked_list2.add_data(6)
linked_list2.add_data(7)
linked_list2.show_data()

linked_list2 = reverse_list(linked_list2, 3, 5)
linked_list2.show_data()

File: Practice!/Week2/LinkedListPrac.cpp
// Practice1
// 단방향 연결 리스트에서 중복 데이터를 찾아 삭제하세요.

// 입출력 예시)
// 입력 연결 리스트: 1, 3, 3, 1, 4, 2, 4, 2
// 결과 연결 리스트: 1, 3, 4, 2

#include <iostream>
#include <unordered_set>
using namespace std;

class Node {
public:
    int data;
    Node* next;

    Node(int data) {
        this->data = data;
        this->next = nullptr;
    }
};

class LinkedList {
public:
    Node* head;

    LinkedList() {
        this->head = nullptr;
    }

    void addData(int data) {
        if (this->head == nullptr) {
            this->head = new Node(data);
        } else {
            Node* cur = this->head;
            while (cur->next != nullptr) {
                cur = cur->next;
            }
            cur->next = new Node(data);
        }
    }

    bool findData(int data) {
        Node* cur = this->head;
        while (cur != nullptr) {
            if (cur->data == data) {
                return true;
            }
            cur = cur->next;
        }
        return false;
    }

    void showData() {
        Node* cur = this->head;
        while (cur != nullptr) {
            cout << cur->data << " ";
            cur = cur->next;
        }
        cout << endl;
    }
};

LinkedList removeDup(LinkedList listBefore) {
    //여기에 코드 작성
    return null;
}

int main() {
    LinkedList linkedList;
    linkedList.addData(1);
    linkedList.addData(3);
    linkedList.addData(3);
    linkedList.addData(1);
    linkedList.addData(4);
    linkedList.addData(2);
    linkedList.addData(4);
    linkedList.addData(2);

    cout << "Before removing duplicates: ";
    linkedList.showData();

    LinkedList result = removeDup(linkedList);

    cout << "After removing duplicates: ";
    result.showData();

    return 0;
}


// Practice2
// Palindrome 연결 리스트
// 추가 자료구조 없이 연결 리스트만으로 풀어보기 Practice1에서 만든 단일연결리스트 사용(내장 LinkedList말고)
// Palindrome: 앞으로 읽어도 뒤로 읽어도 같은 문자열

#include <iostream>
#include <unordered_set>
using namespace std;

class Node {
public:
    int data;
    Node* next;

    Node(int data) {
        this->data = data;
        this->next = nullptr;
    }
};

class LinkedList {
public:
    Node* head;

    LinkedList() {
        this->head = nullptr;
    }

    void addData(int data) {
        if (this->head == nullptr) {
            this->head = new Node(data);
        } else {
            Node* cur = this->head;
            while (cur->next != nullptr) {
                cur = cur->next;
            }
            cur->next = new Node(data);
        }
    }

    bool findData(int data) {
        Node* cur = this->head;
        while (cur != nullptr) {
            if (cur->data == data) {
                return true;
            }
            cur = cur->next;
        }
        return false;
    }

    void showData() {
        Node* cur = this->head;
        while (cur != nullptr) {
            cout << cur->data << " ";
            cur = cur->next;
        }
        cout << endl;
    }
};

bool checkPalindrome(LinkedList& list) {
    //여기에 코드 작
    return true;
}

int main() {
    LinkedList linkedList1;
    linkedList1.addData(1);
    linkedList1.addData(3);
    linkedList1.addData(5);
    linkedList1.addData(3);
    linkedList1.addData(1);
    cout << checkPalindrome(linkedList1) << endl; // true

    LinkedList linkedList2;
    linkedList2.addData(3);
    linkedList2.addData(5);
    linkedList2.addData(5);
    linkedList2.addData(3);
    cout << checkPalindrome(linkedList2) << endl; // true

    LinkedList linkedList3;
    linkedList3.addData(1);
    linkedList3.addData(3);
    linkedList3.addData(5);
    linkedList3.addData(1);
    cout << checkPalindrome(linkedList3) << endl; // false

    return 0;
}




// Practice3
// 연결 리스트 부분 뒤집기
// 주어진 연결 리스트에서 시작 위치부터 끝 위치 사이의 노드들을 뒤집으시오.

// 입력 예시)
// 입력 연결 리스트: 1, 2, 3, 4, 5
// 시작 위치: 2
// 끝 위치: 4
// (처음 위치는 1부터라고 가정)
// 결과 연결 리스트: 1, 4, 3, 2, 5
#include <iostream>
#include <unordered_set>
using namespace std;

class Node {
public:
    int data;
    Node* next;

    Node(int data) {
        this->data = data;
        this->next = nullptr;
    }
};

class LinkedList {
public:
    Node* head;

    LinkedList() {
        this->head = nullptr;
    }

    void addData(int data) {
        if (this->head == nullptr) {
            this->head = new Node(data);
        } else {
            Node* cur = this->head;
            while (cur->next != nullptr) {
                cur = cur->next;
            }
            cur->next = new Node(data);
        }
    }

    bool findData(int data) {
        Node* cur = this->head;
        while (cur != nullptr) {
            if (cur->data == data) {
                return true;
            }
            cur = cur->next;
        }
        return false;
    }

    void showData() {
        Node* cur = this->head;
        while (cur != nullptr) {
            cout << cur->data << " ";
            cur = cur->next;
        }
        cout << endl;
    }
};


    LinkedList reverseList(int left, int right) {
        //여기에 코드 작성
        return *this;
    }
};

int main() {
    LinkedList linkedList;
    linkedList.addData(1);
    linkedList.addData(2);
    linkedList.addData(3);
    linkedList.addData(4);
    linkedList.addData(5);
    linkedList.showData();

    linkedList.reverseList(2, 4);
    linkedList.showData();

    LinkedList linkedList2;
    linkedList2.addData(1);
    linkedList2.addData(2);
    linkedList2.addData(3);
    linkedList2.addData(4);
    linkedList2.addData(5);
    linkedList2.addData(6);
    linkedList2.addData(7);
    linkedList2.showData();

    linkedList2.reverseList(3, 5);
    linkedList2.showData();

    return 0;
}

